readme.md
![](logo.png)

# Howerpoint - forget about Powerpoint!!

* **ghci** terminal-based slide-show presentations that you can control from the REPL and execute code on slides
* idea (no code) stolen from [REPLesent](https://github.com/marconilanna/REPLesent)

_It's only a single Haskell file!_

[![asciicast](https://asciinema.org/a/42096.png)](https://asciinema.org/a/42096)

To run presentation create presentation as a text file.
Run `stack ghci` and have in the in the root of the project

    $ stack ghci
    Howerpoint> :loadPresentation PATH-TO-PRESENTATION

Good stating poitn is test presentations

    $ stack ghci
    Howerpoint> :loadTestPresentation

### Navigation commands
Command | Description
--- | ---
`:loadTestPresentation` | Load test presentation
`:loadPresentation <PATH-TO-YOUR-PRESENTATION>` | Load your presentation
`help` | Help
`n` | Go to next slide
`p` | Go to previous slide
`nn i` | Advance i slides
`pp i` | Go back i slides
`g i` | Go to slide i
`:rr` | Run code from slide
`:rrr` | Run code from current and all previous slides
`resetSize` | Set the slide size to current terminal size
`setSize x y` | Set the slide size to x y
`showSize` | Show current slide size

<!--
Command | Shortcut | Symbolic alias | Description
--- | --- | --- | ---
`next` | `n` | `>` | Go to next build/slide
`previous` | `p` | `<` | Go back to previous build/slide
`redraw` | `z` | | Redraw the current build/slide
`Next` | `N` | `>>` | Go to next slide
`Previous` | `P` | `<<` | Go back to previous slide
`i next` | `i n` | | Advance i slides
`i previous` | `i p` | | Go back i slides
`i go` | `i g` | | Go to slide i
`first` | `f` | `|<` | Go to first slide
`last` | `l` | `>|` | Go to last slide
`Last` | `L` | `>>|` | Go to last build of last slide
`run` | `r` | `!!` | Execute code that appears on slide
`blank` | `b` | | Blank screen
`help` | `h` | `?` | This help message
-->


###ANSI colors

Escape code	| Result
--- | ---
\x | Foreground color, where x is one of: blac`k`, `r`ed, `g`reen, `y`ellow, `b`lue, `m`agenta, `c`yan, `w`hite
\X | Background color, where capital X is one of the same as above
\0 | Resets foreground and background colors to terminal default

###Code

All code is run in ghci so don't forget `let`s and `:m +` as imports etc.

**\>\>\>** show code on a slide and also the same code will be run

    >>>putStrLn . show $ double 5

**L\>\>** show code on a slide and when the code is run it prepends `let`

    L>>double x = x + x

**H\>\>** doesn't show code on a slide but the code is run. It's useful for imports

    H>>:m +Data.Char


changelog.md
# Change Log

## [Unreleased][unreleased]
### Added
- :rrr command

### Changed
- removed Setup.hs


## [v1] - 2016-04-11
### Added
- Initial public release
- Basic navigation
- Color support
- Code running support
- Size auto-detection + setting size of presentation

[unreleased]: https://github.com/fokot/howerpoint/compare/v1...HEAD
[v1]: https://github.com/fokot/howerpoint/tree/v1


src/Main.hs
{-# LANGUAGE ScopedTypeVariables #-}

module Main where

import System.IO
import Control.Monad
-- import Data.List.Split
--  >>= (\x -> return $ splitOn "\n" x)

f = "/Users/Frantisek/haskell/howerpoint/test/resources/test_navigation.txt"

data PresentationSetting = PresentationSetting {
  width :: Int,
  height :: Int
}

padding :: Int -> Int -> (Int, Int)
padding size contentSize = (half, if odd (size - contentSize) then half + 1 else half)
  where half = floor $ (fromIntegral (size - contentSize)) / 2


repeatN n = take n . repeat

drawFrame :: PresentationSetting -> [String] -> [String]
drawFrame ps content =
  line :  (repeatN topPadding emptyLine) ++ map centerLine content ++ repeatN bottomPadding emptyLine ++ [line]
  where (topPadding, bottomPadding) = padding (height ps) (length content)
        line = take (width ps) $ repeat '*'
        emptyLine = '*' : (repeatN ((width ps) -2) ' ') ++ "*"
        centerLine :: String -> String
        centerLine lineContent =
          let (leftPadding, rightPadding) = padding (width ps) (length lineContent)
          in '*' :
             (repeatN (leftPadding - 1) ' ') ++
             lineContent ++
             (repeatN (rightPadding - 1) ' ') ++
             "*"


loadFile :: String -> IO ()
loadFile path = withFile path ReadMode (\handle -> do
                content :: [String] <- liftM lines $ hGetContents handle
--                 let slides :: [[String]]= content
                let frame :: [String] = drawFrame (PresentationSetting 50 20) content
                mapM putStrLn frame
                return ())
main = putStr "hi"

src/Howepoint.hs
{-# LANGUAGE ScopedTypeVariables #-}

module Howepoint where

import System.IO
import Data.List.Split
import Data.IORef
import Control.DeepSeq (deepseq)
import qualified System.Console.Terminal.Size as T
import Data.Char (isLower, toLower)
import Data.Maybe (isJust, fromMaybe)

f = "/Users/Frantisek/haskell/howerpoint/test/resources/howerpoint.txt"

type Slide = [String]

data PresentationState = PresentationState {
  slides :: [Slide],
  width :: Int,
  height :: Int,
  currentSlide :: Int
}

h = do
  putStrLn "To start presentation run:"
  putStrLn "  :set -XRecordWildCards"
  putStrLn "  Presentation {..} <- loadPresentation \"PATH-TO-PRESENTATION\""


data Presentation = Presentation {
  n :: IO (),
  p :: IO (),
  nn :: Int -> IO (),
  pp :: Int -> IO (),
  g :: Int -> IO (),
  resetSize :: IO (),
  setSize :: Int -> Int -> IO (),
  showSize :: IO ()
}

parseSlides :: String -> IO [Slide]
parseSlides path = do
    h <- openFile path ReadMode
    slides <- (splitOn ["---"]) . lines <$> hGetContents h
    slides `deepseq` hClose h
    pure slides

displaySlide :: IORef PresentationState -> IO ()
displaySlide ps = do
  ps <- readIORef ps
  let slide = (slides ps) !! (currentSlide ps)
  mapM putStrLn (formatSlide ps slide) >> pure ()

moveSlide :: IORef PresentationState -> (Int -> Int) -> IO ()
moveSlide ps f = do
  modifyIORef ps (\p -> p {currentSlide = f (currentSlide p)})
  displaySlide ps

getTerminalSize :: IO (Int, Int)
getTerminalSize = T.size >>= (\s -> pure $ maybe (50, 20) (\w -> (T.width w, T.height w)) s)

loadPresentation :: String -> IO Presentation
loadPresentation path = do
  slides <- parseSlides path
  (w, h) <- getTerminalSize
  ioRef <- newIORef $ PresentationState slides w (h - 3) 0
  displaySlide ioRef
  let maxSlides = length slides
  let nextSlide = moveSlide ioRef (\x -> if x == maxSlides - 1 then x else x + 1)
  let previousSlide = moveSlide ioRef (\x -> if x == 0 then x else x - 1)
  let advanceSlides = \i -> moveSlide ioRef (\x -> if x + i >= maxSlides - 1 then maxSlides - 1 else x + i)
  let goBackSlides = \i -> moveSlide ioRef (\x -> if x - i <= 0 then 0 else x - i)
  let goToSlide = \i -> moveSlide ioRef (\x -> if i > 0 && i <= maxSlides then i - 1 else x)
  let resetSize = getTerminalSize >>= (\(w, h) -> modifyIORef ioRef (\p -> p {width = w, height = h - 3} ) ) >> displaySlide ioRef
  let setSize w h = modifyIORef ioRef (\p -> p {width = w, height = h}) >> displaySlide ioRef
  let showSize = readIORef ioRef >>= (\ps -> putStrLn $ (show $ width ps) ++ "x" ++ (show $ height ps) )
  return $ Presentation {
    n = nextSlide,
    p = previousSlide,
    nn = advanceSlides,
    pp = goBackSlides,
    g = goToSlide,
    resetSize = resetSize,
    setSize = setSize,
    showSize = showSize
  }

-- |calculates left-right or top-bottom padding from the maximum size and content size
padding :: Int -> Int -> (Int, Int)
padding size contentSize = (half, if odd (size - contentSize) then half + 1 else half)
  where half = floor $ (fromIntegral (size - contentSize)) / 2

repeatN n = take n . repeat

formatSlide :: PresentationState -> Slide -> Slide
formatSlide ps content =
  line :  (repeatN topPadding emptyLine) ++ map centerLine content ++ repeatN bottomPadding emptyLine ++ [line]
  where (topPadding, bottomPadding) = padding (height ps) (length content)
        line = take (width ps) $ repeat '*'
        emptyLine = '*' : (repeatN ((width ps) -2) ' ') ++ "*"
        centerLine :: String -> String
        centerLine lineContent =
          let (colorizedContent, len) = colorize lineContent
              (leftPadding, rightPadding) = padding (width ps) len
          in '*' :
             (repeatN (leftPadding - 1) ' ') ++
             colorizedContent ++
             (repeatN (rightPadding - 1) ' ') ++
             "*"

-- foreground colors 30-37, background colors 40-47
color :: Char -> Maybe String
color x = fmap (\c -> "\x1b[" ++ show (if isLower x then c + 30 else c + 40) ++ "m" )
  (case toLower x of  'k' -> Just 0 -- Black
                      'r' -> Just 1 -- Red
                      'g' -> Just 2 -- Green
                      'y' -> Just 3 -- Yellow
                      'b' -> Just 4 -- Blue
                      'm' -> Just 5 -- Magenta
                      'c' -> Just 6 -- Cyan
                      'w' -> Just 7 -- White
                      _   -> Nothing -- No coloring
  )

colorReset = "\x1b[0m"

-- it also takes care of escaping to ;\\k' will be just '\k' and not black color
colorize :: String -> (String, Int)
colorize line = if elem '\\' line then colorize' line [] (length line) else (line, length line)
  where colorize' [] acc len = (acc ++ colorReset, len)
        colorize' ('\\':x:xs) acc len = let maybeColor = color x
                                            colorizedChar = fromMaybe [x] maybeColor
                                        in  colorize' xs (acc ++ colorizedChar) (if isJust maybeColor then len - 2 else len)
        colorize' (x:xs) acc len = colorize' xs (acc ++ [x]) len

src/Howerpoint.hs
{-# LANGUAGE RecordWildCards     #-}
{-# LANGUAGE ScopedTypeVariables #-}

module Howerpoint where

import           Control.DeepSeq              (deepseq)
import           Control.Monad                ((>=>), liftM)
import           Data.Char                    (isLower, toLower)
import           Data.IORef
import           Data.List                    (isPrefixOf)
import           Data.List.Split
import           Data.Maybe                   (fromMaybe, isJust, mapMaybe)
import qualified System.Console.Terminal.Size as T
import           System.IO

type Slide = [String]

data Zipper a = Zipper [a] a [a]

extract :: Zipper a -> a
extract (Zipper _ x _) = x

previous :: Zipper a -> [a]
previous (Zipper x _ _) = x

fromList :: [a] -> Zipper a
fromList (x : xs)
  = Zipper [] x xs
fromList []
  = error "Empty list"

moveLeft :: Zipper a -> Maybe (Zipper a)
moveLeft (Zipper (l : ls) x rs)
  = Just $ Zipper ls l (x : rs)
moveLeft z
  = Nothing

moveRight :: Zipper a -> Maybe (Zipper a)
moveRight (Zipper ls x (r : rs))
  = Just $ Zipper (x : ls) r rs
moveRight z
  = Nothing

data PresentationState = PresentationState {
  slides       :: Zipper Slide,
  width        :: Int,
  height       :: Int,
  currentSlide :: Int
}

getCurrentSlide :: PresentationState -> Slide
getCurrentSlide = extract . slides

getPreviousSlides :: PresentationState -> [Slide]
getPreviousSlides = previous . slides

sanitizeCode :: String -> String
sanitizeCode = concatMap (\c -> if c == '"' then "\\\"" else [c])

filterCode :: Slide -> [String]
filterCode = mapMaybe processLine
  where
    processLine :: String -> Maybe String
    processLine line
      | ">>>" `isPrefixOf` line = Just $ drop 3 line
      | "L>>" `isPrefixOf` line = Just $ "let " ++ drop 3 line
      | "H>>" `isPrefixOf` line = Just $ drop 3 line
      | otherwise               = Nothing

isCode :: String -> Bool
isCode s = isPrefixOf ">>>" s || isPrefixOf "L>>" s

isHiddenCode :: String -> Bool
isHiddenCode = isPrefixOf "H>>"

maxCodeLength :: Slide -> Int
maxCodeLength = maximum . map (subtract 3 . length) . filter isCode

printCodeAndExecute :: [String] -> String
printCodeAndExecute code = unlines (map (\x -> "putStrLn \"> " ++ sanitizeCode x ++ "\"") code) ++ unlines code

help :: IO ()
help = mapM_ (putStrLn . fst . colorize)
       [ "\\k\\WTo load test presentation run \\Y:loadTestPresentation"
       , "\\k\\WTo load your own presentation run \\Y:loadPresentation <PATH-TO-YOUR-PRESENTATION>"]


data Presentation = Presentation {
  n                      :: IO (),
  p                      :: IO (),
  nn                     :: Int -> IO (),
  pp                     :: Int -> IO (),
  g                      :: Int -> IO (),
  resetSize              :: IO (),
  setSize                :: Int -> Int -> IO (),
  showSize               :: IO (),
  codeFromSlide          :: IO [String],
  codeFromPreviousSlides :: IO [[String]]
}

parseSlides :: String -> IO (Zipper Slide)
parseSlides path = do
    h <- openFile path ReadMode
    slides' <- splitOn ["---"] . lines <$> hGetContents h
    slides' `deepseq` hClose h
    return $ fromList slides'

displaySlide :: IORef PresentationState -> IO ()
displaySlide ps = do
  ps' <- readIORef ps
  mapM_ putStrLn (formatSlide ps' (getCurrentSlide ps'))

moveSlide :: (Int -> Int) -> PresentationState -> PresentationState
moveSlide f ps
  = maybe ps (\ss -> ps {slides = ss, currentSlide = newpos}) slides'
  where ntimes n fun = foldr (>=>) return (replicate n fun)
        pos = currentSlide ps
        newpos = f pos
        dir = if pos < newpos
              then moveRight
              else moveLeft
        slides' = ntimes (abs $ newpos - pos) dir (slides ps)

getTerminalSize :: IO (Int, Int)
getTerminalSize = do
  s <- T.size
  return $ maybe (50, 20) (\w -> (T.width w, T.height w)) s

loadPresentation :: String -> IO Presentation
loadPresentation path = do
  slides' <- parseSlides path
  (w, h) <- getTerminalSize
  ioRef <- newIORef $ PresentationState slides' w (h - 3) 0

  displaySlide ioRef
  let
    g  = moveWith . const
    nn = moveWith . (+)
    pp = moveWith . subtract
    p  = pp 1
    n  = nn 1

    moveWith f
      = modifyIORef ioRef (moveSlide f) >> displaySlide ioRef

    resetSize = do
      (w', h') <- getTerminalSize
      setSize w' (h' - 3)

    setSize w' h' = do
      modifyIORef ioRef (\p' -> p' {width = w', height = h'})
      displaySlide ioRef

    showSize = do
      ps <- readIORef ioRef
      putStrLn $ show (width ps) ++ "x" ++ show (height ps)

    codeFromSlide = filterCode . getCurrentSlide <$> readIORef ioRef

    codeFromPreviousSlides = liftM (map filterCode) $ getPreviousSlides <$> readIORef ioRef

  return Presentation {..}

-- |calculates left-right or top-bottom padding from the maximum size and content size
padding :: Int -> Int -> (Int, Int)
padding size contentSize = (half, if odd (size - contentSize) then half + 1 else half)
  where half = (size - contentSize) `div` 2

formatSlide :: PresentationState -> Slide -> Slide
formatSlide ps content =
  lineOfStars : replicate topPadding emptyLine ++ map centerLine showableContent ++ replicate bottomPadding emptyLine ++ [lineOfStars]
  where showableContent = filter (not . isHiddenCode) content
        (topPadding, bottomPadding) = padding (height ps) (length showableContent)
        lineOfStars = replicate (width ps) '*'
        emptyLine = '*' : replicate (width ps - 2) ' ' ++ "*"
        (leftCodePadding, _) = padding (width ps) (maxCodeLength showableContent - 3)
        trimCode :: String -> String
        trimCode = drop 3
        centerLine :: String -> String
        centerLine lineContent =
          let (colorizedContent, len) =
                if isCode lineContent
                    then colorize (trimCode lineContent)
                    else colorize lineContent
              (leftPadding, rightPadding) =
                if isCode lineContent
                  then (leftCodePadding, width ps - leftCodePadding - len)
                  else padding (width ps) len
          in '*' :
             replicate (leftPadding - 1) ' ' ++
             colorizedContent ++
             replicate (rightPadding - 1) ' ' ++
             "*"

-- foreground colors 30-37, background colors 40-47
color :: Char -> Maybe String
color '0'
  = Just colorReset
color x
  = fmap (\c -> "\x1b[" ++ show (if isLower x then c + 30 else c + 40) ++ "m")
         (lookup (toLower x) colors)
  where
    colors :: [(Char, Int)]
    colors =
      [ ('k', 0) -- Black
      , ('r', 1) -- Red
      , ('g', 2) -- Green
      , ('y', 3) -- Yellow
      , ('b', 4) -- Blue
      , ('m', 5) -- Magenta
      , ('c', 6) -- Cyan
      , ('w', 7) -- White
      ]

colorReset :: String
colorReset = "\x1b[0m"

-- it also takes care of escaping to '\\k' will be just '\k' and not black color
colorize :: String -> (String, Int)
colorize line =
  if '\\' `elem` line
    then colorize' line [] (length line)
    else (line, length line)
  where colorize' [] acc len = (acc ++ colorReset, len)
        colorize' ('\\':x:xs) acc len =
          let maybeColor = color x
              colorizedChar = fromMaybe [x] maybeColor
              len' = if isJust maybeColor then len - 2 else len
          in  colorize' xs (acc ++ colorizedChar) len'
        colorize' (x:xs) acc len = colorize' xs (acc ++ [x]) len
